sys/memarray/memarray.c

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (C) 2018 Tobias Heider <heidert@nm.ifi.lmu.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+#include <string.h>
+#include "memarray.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+void memarray_init(memarray_t *mem, void *data, size_t size, size_t num)
+{
+    assert((mem != NULL) && (data != NULL) && (size >= sizeof(void *)));
+
+    DEBUG("memarray: Initialize memarray of %u times %u Bytes at %p\n",
+          (unsigned)num, (unsigned)size, data);
+
+    mem->free_data = data;
+    mem->size = size;
+    mem->num = num;
+
+    for (size_t i = 0; i < (mem->num - 1); i++) {
```

Comment:
This makes ```mem->num - 1``` re-evaluate on every loop iteration.

Reply:
Does it ? has `mem->num` is not volatile there is no reason for the compiler to re-read it I think.
When replacing with an intermediate variable I get the same size for `samr21-xpro`, `wsn430-v1_3b` and `native`.
If you have another solution that reduces code size it would be good.

When using `(num -1)` in the loop it gets bigger for `samr21-xpro`, no idea why.

Reply:
Any function call in between two dereferences usually forces re-read from memory. The called function might have changed what's behind that pointer, after all. But maybe the compiler has outsmarted me (again).

Reply:
Yeah `memcpy` could be doing some stuff to `mem->num` by overwriting it if `mem` and `data` overlap and in that case the compiler should indeed re-read it. Maybe it is re-read but does not cost more.


I found one that reduces code size and also solves the `num == 0` case.
But it would print many debug free message at init so maybe not that good.

```
    for (char *cur = data; cur < (char *)data + (num * size); cur += size) {
        memarray_free(mem, cur);
    }
```
We can keep the optimizations to another PR this way it is not tied to tobhe.